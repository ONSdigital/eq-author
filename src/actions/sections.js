export const ADD_SECTION = 'ADD_SECTION';
export const ADD_QUESTION = 'ADD_QUESTION';
export const ADD_ANSWER = 'ADD_ANSWER';

export const addSection = name => {
  return {
    type: ADD_SECTION,
    payload: {
      name: name,
    },
  };
};

export const addQuestion = (name, sectionID) => {
  return {
    type: ADD_QUESTION,
    payload: {
      name: name,
      sectionID: sectionID,
    },
  };
};

export const addAnswer = (name, questionID, sectionID) => {
  return {
    type: ADD_ANSWER,
    payload: {
      name: name,
      questionID: questionID,
      sectionID: sectionID,
    },
  };
};
