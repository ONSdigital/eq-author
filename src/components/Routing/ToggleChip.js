import React from "react";
import styled from "styled-components";

import { colors } from "constants/theme";
import closeIcon from "./icon-chip-unselected.svg";
import checkIcon from "./icon-chip-check.svg";

const Field = styled.div`
  position: relative;
`;

const Input = styled.input`
  background: url('${props =>
    props.defaultChecked ? checkIcon : closeIcon}') no-repeat 0 0;
  width: 1.5em;
  height: 1.5em;
  border: none;
  position: absolute;
  left: 1em;
  top: 0;
  bottom: 0;
  margin: auto;
  -webkit-appearance: none;
  pointer-events: none;
  &:focus {
    opacity: 1;
    outline: none;
  }
`;

const Label = styled.label`
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  background: ${props => (props.checked ? "#0288D1" : colors.borders)};
  color: ${props => (props.checked ? colors.white : colors.text)};
  padding: 0.5em 1.5em 0.5em 2.5em;
  font-size: 0.9em;
  margin: 0.2em;
  border-radius: 1em;
  display: flex;
  align-items: center;
  position: relative;
  user-select: none;
  cursor: pointer;
  transition: all 200ms;
  &:hover {
    background: ${props => (props.checked ? "#0479b9" : "#dadada")};
  }
`;

const ToggleChip = ({ label, id, name, checked, onChange }) => (
  <Field>
    <Label checked={checked} htmlFor={id}>
      {label}
    </Label>
    <Input
      id={id}
      name={id}
      defaultChecked={checked}
      type="checkbox"
      onChange={function(e) {
        onChange(name, e.target.checked);
      }}
    />
  </Field>
);

export default ToggleChip;
