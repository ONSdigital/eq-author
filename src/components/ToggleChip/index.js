import React from "react";
import PropTypes from "prop-types";
import styled, { css } from "styled-components";
import { colors } from "constants/theme";
import checkedIcon from "./checked.svg";

const labelStyles = {
  checked: css`
    color: ${colors.white};
    background: #008dd0;
    border-color: #008dd0;

    &:hover {
      background: #007ab3;
    }
  `,
  unchecked: css`
    color: ${colors.text};
    background: #f8f8f8;
    border-color: rgba(0, 141, 208, 0);

    &:hover {
      background: #eeeded;
    }
  `
};

const focusStyles = {
  checked: css`
    color: ${colors.white};
    background: #008dd0;
    border-color: #005e8b;
  `,
  unchecked: css`
    color: ${colors.text};
    background: #eeeded;
    border-color: #008dd0;
  `
};

const checkboxStyle = {
  checked: css`
    border-color: white;
    background: white url(${checkedIcon}) no-repeat center;
    background-size: contain;
  `,
  unchecked: css`
    background-color: rgba(255, 255, 255, 0);
    border-color: #979797;
  `
};

const Field = styled.div`
  display: inline-block;
  position: relative;
  margin-right: 0.2em;
  margin-bottom: 0.2em;
`;

export const Label = styled.label`
  display: flex;
  align-items: center;
  user-select: none;
  border-radius: 2.75em;
  font-size: 0.8em;
  padding: 0.5em 0.8em 0.5em 2.5em;
  cursor: pointer;
  border: 2px solid;
  ${props => (props.checked ? labelStyles.checked : labelStyles.unchecked)};
`;

const Truncated = styled.span`
  display: inline-block;
  max-width: ${props => props.maxWidth}em;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

Truncated.defaultProps = {
  maxWidth: 16
};

export const Input = styled.input`
  width: 1.2em;
  height: 1.2em;
  position: absolute;
  left: 0.8em;
  top: 0;
  bottom: 0;
  margin: auto;
  -webkit-appearance: none;
  pointer-events: none;
  z-index: 2;
  border-radius: 50%;
  border: 2px solid;
  ${props => (props.checked ? checkboxStyle.checked : checkboxStyle.unchecked)};

  &:focus {
    opacity: 1;
    outline: none;
  }

  /* stylelint-disable */
  &:focus + ${Label} {
    /* stylelint-enable */
    ${props => (props.checked ? focusStyles.checked : focusStyles.unchecked)};
  }
`;

class ToggleChip extends React.Component {
  static propTypes = {
    id: PropTypes.string,
    title: PropTypes.string,
    children: PropTypes.node,
    checked: PropTypes.bool.isRequired,
    maxWidth: PropTypes.number,
    onChange: PropTypes.func
  };

  handleToggle = () => {
    this.props.onChange({
      name: this.props.id,
      value: !this.props.checked
    });
  };

  render() {
    const { id, title, children, checked, maxWidth } = this.props;
    return (
      <Field>
        <Input
          id={id}
          name={id}
          type="checkbox"
          checked={checked}
          onChange={this.handleToggle}
        />
        <Label htmlFor={id} checked={checked}>
          <Truncated title={title} maxWidth={maxWidth}>
            {children}
          </Truncated>
        </Label>
      </Field>
    );
  }
}

export default ToggleChip;
