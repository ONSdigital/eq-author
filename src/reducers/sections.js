import _, {merge} from 'lodash';
import * as actions from 'actions/sections';

const ids = [];
const getIdFromName = name => {
  const newId = name.replace(/\s+/g, '-').toLowerCase();
  ids.push(newId);
  return newId;
};

const getSectionById = (sections, id) =>
  _(sections).filter({id: id}).value()[0];

const sections = (state = {sections: []}, action) => {
  switch (action.type) {
    case actions.ADD_SECTION:
      return {
        sections: [
          ...state.sections,
          {
            id: getIdFromName(action.payload.name),
            name: action.payload.name,
            questions: [],
          },
        ],
      };

    case actions.ADD_QUESTION:
      return merge(
        {},
        state.sections,
        getSectionById(
          state.sections,
          action.payload.sectionID
        ).questions.push({
          id: getIdFromName(action.payload.name),
          name: action.payload.name,
          answers: [],
        })
      );

    case actions.ADD_ANSWER:
      _(state.sections)
        .map('questions')
        .flatten()
        .filter({id: action.payload.questionID})
        .value()[0]
        .answers.push({
          id: getIdFromName(action.payload.name),
          name: action.payload.name,
        });

      return merge(
        {},
        state.sections,
        getSectionById(state.sections, action.payload.sectionID)
      );

    default:
      return state;
  }
};

export default sections;
