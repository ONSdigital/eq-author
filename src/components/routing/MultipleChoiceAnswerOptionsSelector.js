import React from "react";
import styled from "styled-components";

import ToggleChip from "components/ToggleChip";

import { PropTypes } from "prop-types";

const MultipleChoiceAnswerOptions = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  padding: 1em 0;
`;

const MultipleChoiceAnswerOptionsSelector = ({
  options,
  onOptionSelectionChange
}) => (
  <MultipleChoiceAnswerOptions>
    {options.map(option => (
      <ToggleChip
        key={option.id}
        id={option.id}
        checked={option.selected}
        title={option.label}
        onChange={onOptionSelectionChange}
      >
        {option.label || <strong>Unlabelled option</strong>}
      </ToggleChip>
    ))}
  </MultipleChoiceAnswerOptions>
);

MultipleChoiceAnswerOptionsSelector.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      selected: PropTypes.bool.isRequired,
      label: PropTypes.string.isRequired
    })
  ).isRequired,
  onOptionSelectionChange: PropTypes.func.isRequired
};

export default MultipleChoiceAnswerOptionsSelector;
