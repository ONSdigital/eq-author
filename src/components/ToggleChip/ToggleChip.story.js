import React from "react";
import { storiesOf } from "@storybook/react";
import styled from "styled-components";
import { ToggleChip, ToggleChipGroup } from "components/ToggleChip";
import { uniqueId } from "lodash";
import { h1 } from "draft-js-raw-content-state/build/utils/blockTypes";

const Wrapper = styled.div`
  padding: 1em;
  width: 50%;
`;

class StatefulToggleChip extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: false
    };
  }

  handleChange = e => {
    this.setState({
      value: e.target.checked
    });
  };

  render() {
    return (
      <ToggleChip
        id={uniqueId()}
        onChange={this.handleChange}
        value={uniqueId()}
        checked={this.state.value}
        {...this.props}
      />
    );
  }
}

console.log(ToggleChipGroup);

class ToggleChipGroupContainer extends React.Component {
  state = {
    value: "123"
  };

  handleChange = ({ value }) => {
    this.setState({ value });
  };

  render() {
    return (
      <ToggleChipGroup
        onChange={this.handleChange}
        value={this.state.value}
        name="haha"
      >
        <ToggleChip title="hello" value="1">
          foo
        </ToggleChip>
        <ToggleChip title="hello" value="2">
          bar
        </ToggleChip>
        <ToggleChip title="hello" value="123">
          blah
        </ToggleChip>
      </ToggleChipGroup>
    );
  }
}

storiesOf("ToggleChip", module)
  .add("Single", () => (
    <Wrapper>
      <StatefulToggleChip title="Lorem ipsum dolor sit amet.">
        Lorem ipsum dolor sit amet.
      </StatefulToggleChip>
    </Wrapper>
  ))
  .add("Multiple", () => (
    <Wrapper>
      <StatefulToggleChip title="Lorem ipsum dolor.">
        Lorem ipsum dolor.
      </StatefulToggleChip>
      <StatefulToggleChip title="Lorem ipsum dolor sit.">
        Lorem ipsum dolor sit.
      </StatefulToggleChip>
      <StatefulToggleChip title="Lorem ipsum.">Lorem ipsum.</StatefulToggleChip>
      <StatefulToggleChip title="Lorem ipsum dolor sit amet.">
        Lorem ipsum dolor sit amet.
      </StatefulToggleChip>
      <StatefulToggleChip title="Lorem ipsum dolor sit.">
        Lorem ipsum dolor sit.
      </StatefulToggleChip>
      <StatefulToggleChip title="Lorem ipsum.">Lorem ipsum.</StatefulToggleChip>
      <StatefulToggleChip title="Lorem.">Lorem.</StatefulToggleChip>
      <StatefulToggleChip title="Lorem ipsum dolor sit amet, consectetur adipisicing elit. Earum quos
      saepe voluptatem. Ab consectetur dolore ea eaque excepturi praesentium
      voluptate.">
        Lorem ipsum dolor sit amet, consectetur adipisicing elit. Earum quos
        saepe voluptatem. Ab consectetur dolore ea eaque excepturi praesentium
        voluptate.
      </StatefulToggleChip>
    </Wrapper>
  ))
  .add("Grouped", () => <ToggleChipGroupContainer />);
