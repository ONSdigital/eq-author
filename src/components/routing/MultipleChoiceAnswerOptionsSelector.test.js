import React from "react";
import { shallow } from "enzyme";

import MultipleChoiceAnswerOptionsSelector from "./MultipleChoiceAnswerOptionsSelector";
import sections from "./mockstate";
import ToggleChip from "components/ToggleChip";

let wrapper, mockHandlers;

describe("components/MultipleChoiceAnswerOptionsSelector", () => {
  beforeEach(() => {
    mockHandlers = {
      onSelectAll: jest.fn(),
      onOptionSelectionChange: jest.fn()
    };

    wrapper = shallow(
      <MultipleChoiceAnswerOptionsSelector
        options={sections[0].pages[0].options}
        {...mockHandlers}
      />
    );
  });

  it("should render consistently", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should handle change events on ToggleChip", () => {
    wrapper
      .find(ToggleChip)
      .first()
      .simulate("change");
    expect(mockHandlers.onOptionSelectionChange).toHaveBeenCalled();
  });

  it("should handle 'select all' button", () => {
    wrapper.find("[data-test='btn-select-all']").simulate("click");
    expect(mockHandlers.onSelectAll).toHaveBeenCalled();
  });
});
