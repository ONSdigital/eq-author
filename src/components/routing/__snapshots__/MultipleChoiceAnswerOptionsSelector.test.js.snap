// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/MultipleChoiceAnswerOptionsSelector should render consistently 1`] = `
<MultipleChoiceAnswerOptionsSelector__MultipleChoiceAnswerOptions
  data-test="options-selector"
>
  <ToggleChipGroup
    onChange={[Function]}
    value="2"
  >
    <ToggleChip
      key="1"
      title="a"
      value="1"
    >
      a
    </ToggleChip>
    <ToggleChip
      key="2"
      title="b"
      value="2"
    >
      b
    </ToggleChip>
    <ToggleChip
      key="3"
      title="c"
      value="3"
    >
      c
    </ToggleChip>
  </ToggleChipGroup>
</MultipleChoiceAnswerOptionsSelector__MultipleChoiceAnswerOptions>
`;
